orchestrator/careplanservice/policy/agent.go

```diff
@@ -0,0 +1,199 @@
+)
+
+func subjectToParams(subject *fhir.Reference) (url.Values, error) {
+	params := url.Values{}
+
+	if coolfhir.IsLogicalReference(subject) {
+		params["patient:Patient.identifier"] = []string{fmt.Sprintf("%s|%s", *subject.Identifier.System, *subject.Identifier.Value)}
+	} else if subject.Id != nil {
+		params["subject"] = []string{fmt.Sprintf("Patient/%s", *subject.Id)}
+	} else if subject.Reference != nil {
+		params["subject"] = []string{*subject.Reference}
+	} else {
+		return params, fmt.Errorf("invalid subject (subject=%+v)", subject)
+	}
+
+	return params, nil
+}
+
+type SearchCache map[*fhir.Reference][]fhir.CarePlan
+
+func NewSearchCache() SearchCache {
+	return SearchCache{}
+}
+
+//go:embed policy.rego
+var source string
+
+var BuiltinModule = RegoModule{
+	Package: "policy",
+	Source:  source,
+}
+
+var ErrAccessDenied = errors.New("request denied by policy")
+
+type Agent struct {
+	logger zerolog.Logger
+	query  rego.PreparedEvalQuery
+	client fhirclient.Client
+}
+
+type RegoModule struct {
+	Package string
+	Source  string
+}
+
+type Context struct {
+	Principal fhir.Identifier `json:"principal"`
+	Method    string          `json:"method"`
+	Roles     []string        `json:"roles"`
+	Resource  any             `json:"resource"`
+	CarePlans []fhir.CarePlan `json:"careplans"`
+}
+
+func NewAgent(ctx context.Context, module RegoModule, client fhirclient.Client) (Agent, error) {
+	r := rego.New(
+		rego.Query(fmt.Sprintf("allow = data.%s.allow", module.Package)),
+		rego.Module(module.Package, module.Source),
+	)
+	query, err := r.PrepareForEval(ctx)
+	if err != nil {
+		return Agent{}, err
+	}
+
+	return Agent{
+		query:  query,
+		client: client,
+		logger: zerolog.New(os.Stdout),
+	}, nil
+}
+
+func (m Agent) Allow(ctx context.Context, context *Context) error {
+	result, err := m.query.Eval(ctx, rego.EvalInput(context))
+	if err != nil {
+		return fmt.Errorf("failed to evaluate policy: %w", err)
+	}
+
+	if len(result) == 0 {
+		return fmt.Errorf("policy evaluation failed")
+	}
+
+	allow, ok := result[0].Bindings["allow"]
+	if !ok {
+		return fmt.Errorf("allow binding not found")
+	}
+
+	allowed, ok := allow.(bool)
+	if !ok {
+		return fmt.Errorf("invalid type for allow")
+	}
+
+	if !allowed {
+		return ErrAccessDenied
+	}
+
+	return nil
+}
+
+type Preflight struct {
+	ResourceId   string
+	Query        url.Values
+	ResourceType string
+	Principal    fhir.Identifier
+	Method       string
+	Roles        []string
+}
+
+func (a Agent) Preflight(resourceType, id string, r *http.Request) (*Preflight, error) {
+	principal, err := auth.PrincipalFromContext(r.Context())
+	if err != nil {
+		return nil, fmt.Errorf("failed to extract principal from context: %w", err)
+	}
+
+	var roles []string
+
+	rolesHeader := r.Header.Get("Orca-Auth-Roles")
+	if rolesHeader != "" {
+		roles = strings.Split(rolesHeader, ",")
+	}
+
+	return &Preflight{
+		ResourceId:   id,
+		ResourceType: resourceType,
+		Query:        r.URL.Query(),
```

Comment:
We only support POST-based FHIR searches to avoid leaking PII in logging, then the search parameters need to be read from the HTTP request body.

But, you can do a conditional update/create in FHIR in which you _do_ specify search parameters in the HTTP request URL, but we might've removed support for that recently (@dwelman ?); e.g.:

```http
PUT /Task?ehr.identifier=1234
{"resourceType": "Task"}
```

(updates the Task identified by this logical identifier)